### ISSUE TITLE
api: python client library

### ISSUE THREAD

**Issue by lnielsen** [zenodo/zenodo#334]
Create a Python client library for Zenodo API.

**Comment by moble** [comment]
I've gotten a pretty reasonable start on this sort of library [here](https://github.com/moble/sxs/tree/master/sxs/zenodo/api).  I've had good results making new depositions, searching, etc.  You can see an example of a pretty complicated function that uses this interface in the directory right above it that does all sorts of gathering of data, comparing checksums, etc.  Basically, you create a `Login` object which holds the credentials for communicating with the site, then you can create a `Deposit` object that does basically all the tasks the API documentation mentions.

There's also [this](https://github.com/SiLeBAT/zenodo-python) barebones client.

**Comment by abubelinha** [comment]
Aren't these Zenodo Python client API examples?
https://developers.zenodo.org/#quickstart-upload

Sorry if this is a stupid question, but I am quite a newbie in both Zenodo and Python.
Are you talking about some easier stuff, so that you don't need to make the request calls and process responses yourself?
(so there should be a ready-made Python function to do those things for you)

**Comment by moble** [comment]
@abubelinha That's right.  The problem is that those examples quickly get tiring, with lots of repetition.  Pretty much every time you access the API, you have to do something like this:

```python
url = "https://zenodo.org/api/deposit/depositions/1234?access_token=ACCESS_TOKEN"
headers = {"Content-Type": "application/json"}
r = requests.put(url, data=json.dumps(data), headers=headers)
```

You have to construct the whole URL, add your access token, include the headers, and very carefully construct whatever data you're sending.  Then, you might have to deal with weird errors that may just be temporary because the internet connection isn't so great, or the zenodo servers aren't responding just this second.

A library will do a lot of that stuff for you.  For example, the library I mentioned lets you create [a Login object](https://github.com/sxs-collaboration/sxs/blob/b1c607b808d0bf5dd8dd10edd36ffc3eae62a748/sxs/zenodo/api/login.py#L9) that remembers your access token.  Then, you send all your requests through that object, which helps to construct the URL, checks to make sure that it's only sending your token to zenodo, sends the token along with any other relevant headers, and will retry a few times if the request fails for some temporary reason.

Similarly, a library will have easy-to-remember names for things like creating new deposits, uploading or downloading files, etc., so you don't have to look up obscure API details every time you want to do such a thing.

Basically, the point is to make using the API easier and more reliable.

**Comment by abubelinha** [comment]
OK I understand. 
Thanks a lot for your explanations @moble !

I was going to fork the library you mention, but looks like it is part of a much bigger "[sxs](https://github.com/sxs-collaboration/sxs)" project.
Is the stuff in the upper folders required for connecting to Zenodo, or can I just get the stuff under [sxs/zenodo](url) folder? (which [might not be that easy](https://stackoverflow.com/questions/7106012/download-a-single-folder-or-directory-from-a-github-repo), BTW)

Which script should I launch to see a running example?

**Comment by moble** [comment]
I only pointed to that as an example of a pretty extensive API; it's not particularly designed for use by anyone but me.  But if you want to try it, you could indeed just use the stuff in the zenodo module.  There's not really any script that you could use, but you could do something like this:

```python
import sxs.zenodo as zen

# See the documentation for where to put your access token or how to use a password.
# Use the sandbox option when just testing things out.
l = zen.Login(sandbox=True)

# Start a new deposit
d = l.new_deposit

# Below are the required fields for any deposit; for more, see API documentation.
metadata = dict(
    title = "My cool dataset",
    creators = [{"name": "Awesome P. Name"},],
    upload_type = "dataset",
    description = "This dataset describes ...",
)

# Tell zenodo about the metadata.
d.update_metadata(metadata)

# Upload at least one file (or more, if desired).
d.upload_file("/path/to/my/file.txt")

# Tell Zenodo to publish it.
d.publish()

# All done.  Visit the new record here.
d.links["latest_html"]
```

**Comment by abubelinha** [comment]
Thanks a lot indeed for your sample code, @moble 
I preferred to first try your link to the barebones client, where I saw you had also [contributed with code](https://github.com/SiLeBAT/zenodo-python/commit/3f45f9af9031743549a6ad8ebec2d62f32e0431e#diff-141328f6172e960d190e4b5520536e79cac04263898e642e16e038a4dda726ef).

It sort of worked for me, but I would like you to take a look to [this issue](https://github.com/SiLeBAT/zenodo-python/issues/6) I found, where I compare the [file uploading line](https://github.com/SiLeBAT/zenodo-python/blob/master/zenodolib.py#L155) against other Zenodo-Python projects (including your [sxs/zenodo uploading code](https://github.com/sxs-collaboration/sxs/blob/4a4ae0fba945e8de43768dc016319189fff2d496/sxs/zenodo/api/deposit.py#L682)).

But if the bug I suggest is true, I wonder how it worked for you (and other contributors after you).
Maybe the Zenodo-API changed again after all those code changes?

I am also a github newcomer so I don't really know how to contribute [this line](https://github.com/SiLeBAT/zenodo-python/blob/b4ebc5eae890edb09bbce7ccc1c5af6d983790b6/zenodolib.py#L155) change:

`        data = {'file': open(file_path, 'rb')}
`

... should be ...

`        data = open(file_path, 'rb')
`

Could you make the change yourself if you think I am correct?

**Comment by moble** [comment]
My guess is that you're probably right that it shouldn't create a dict — though to be honest, I really don't remember anything about that code.  Also, slint probably has the better approach of using a [context manager](https://www.geeksforgeeks.org/context-manager-in-python/).

But since it's not my project, I can't make the changes myself.  You should probably just open a PR to suggest the changes and let that project's maintainers take a look.  Just go to [that file's github page](https://github.com/SiLeBAT/zenodo-python/blob/master/zenodolib.py), click the little pencil icon in the gray bar right above the file, make your changes, and hit "propose changes".

**Comment by abubelinha** [comment]
I've done that. 
Thanks a lot for the explanations @moble !
